official/utils/logging/logger.py

```diff
@@ -72,3 +84,106 @@ def log_metric(self, name, value, unit=None, global_step=None, extras=None):
       except (TypeError, ValueError) as e:
         tf.logging.warning("Failed to dump metric to log file: "
                            "name %s, value %s, error %s", name, value, e)
+
+  def log_run_info(self, model_name):
+    """Collect most of the TF runtime information for the local env.
+
+    The schema of the run info follows official/benchmark/datastore/schema.
+    """
+    run_info = {"model_name": model_name}
+    _collect_tensorflow_info(run_info)
+    _collect_environment_variable(run_info)
+    _collect_cpu_info(run_info)
+    _collect_gpu_info(run_info)
+    _collect_memory_info(run_info)
+
+    with tf.gfile.GFile(os.path.join(
+        self._logging_dir, _BENCHMARK_RUN_LOG_FILE_NAME), "w") as f:
+      try:
+        json.dump(run_info, f)
+        f.write("\n")
+      except (TypeError, ValueError) as e:
+        tf.logging.warning("Failed to dump benchmark run info to log file: %s",
+                           e)
+
+def _collect_tensorflow_info(run_info):
+  run_info["tensorflow_version"] = {
+      "version": tf.VERSION, "git_hash": tf.GIT_VERSION}
+
+def _collect_environment_variable(run_info):
+  run_info["environment_variable"] = {
+      k:v for k, v in os.environ.items() if k.startswith("TF_")}
+
+# The following code is mirrored from tensorflow/tools/test/system_info_lib
+# which is not exposed for import.
+def _collect_cpu_info(run_info):
+  cpu_info = {}
+
+  cpu_info["num_cores"] = multiprocessing.cpu_count()
+  # Gather num_cores_allowed
+  try:
+    with tf.gfile.GFile("/proc/self/status", "rb") as fh:
+      nc = re.search(r"(?m)^Cpus_allowed:\s*(.*)$", fh.read())
+    if nc:  # e.g. "ff" => 8, "fff" => 12
+      cpu_info["num_cores_allowed"] = (
+        bin(int(nc.group(1).replace(",", ""), 16)).count("1"))
```

Comment:
We should minimally comment this, even if (especially if?) c&ped from TF.

Reply:
Done.